### PR TITLE
fix bug of OrmUtils.compare2Objects

### PR DESCRIPTION
**Pull request by yoshiaki-natsume**
when id is binary and relation is nullable, compare2Objects raise error.

>TypeError: Argument must be a Buffer
>    at Buffer.equals (buffer.js:502:11)
>    at Function.OrmUtils.compare2Objects (../src/util/OrmUtils.ts:176:47)
>    at Function.OrmUtils.compare2Objects (../src/util/OrmUtils.ts:232:31)
>    at Function.OrmUtils.deepCompare (../src/util/OrmUtils.ts:114:23)
>    at Function.EntityMetadata.compareIds (../src/metadata/EntityMetadata.ts:723:25)
>    at ../src/persistence/SubjectChangedColumnsComputer.ts:150:13
>    at Array.forEach (native)
>    at SubjectChangedColumnsComputer.computeDiffRelationalColumns (../src/persistence/SubjectChangedColumnsComputer.ts:119:51)
>    at ../persistence/SubjectChangedColumnsComputer.ts:21:18
>    at Array.forEach (native)
>    at SubjectChangedColumnsComputer.compute (../src/persistence/SubjectChangedColumnsComputer.ts:19:18)
>    at SubjectExecutor.recompute (../src/persistence/SubjectExecutor.ts:167:45)
>    at new SubjectExecutor (../src/persistence/SubjectExecutor.ts:78:14)
>    at EntityPersistExecutor.<anonymous> (../src/persistence/EntityPersistExecutor.ts:119:28)
>    at step (../node_modules/typeorm/persistence/EntityPersistExecutor.js:32:23)
>    at Object.next (../node_modules/typeorm/persistence/EntityPersistExecutor.js:13:53)

this is an error case.

```TypeScript
class A {
  @ManyToOne(type => B, B => B.a, {})
  @JoinColumn({ name: "bId" })
  b: B | null;
}

class B {
  @Column("binary", {name: "id"})
  id: Buffer;

  @OneToMany(type => A, A => A.b)
  a: A[];
}
```

```TypeScript
const a = new A();
b.id = new Buffer([0]);
await a.save();

const b = new B();
await b.save();

a.b = b;
await a.save(); // this raise error
```

### CODE REVIEW

File: src/util/OrmUtils.ts [hunk 1]
```diff
@@ -177,7 +177,9 @@ export class OrmUtils {
         if (x === y)
             return true;
 
-        if (x.equals instanceof Function && x.equals(y))
+        // Check if both argument is Buffer
+        // Buffer.equals Argument must be a Buffer
+        if (Buffer.isBuffer(x) && Buffer.isBuffer(y) && x.equals(y))
```
Comment (yoshiaki-natsume) [inline src/util/OrmUtils.ts]: Does anyone know x and y have `.equals` method expect Buffer?

Comment (pleerock) [reply src/util/OrmUtils.ts]: isn't `.equals` come from buffer? If yes then maybe we can check if x and y are both instances of objects?

Comment (yoshiaki-natsume) [reply src/util/OrmUtils.ts]: equals come from Buffer.
Object.equals is not defined
```TypeScript
const a = new Buffer([]);
const b = {}
a.equals(b); // TypeError: Argument must be a Buffer
b.equals // undefined
```
So you must check instances are both Buffer or not.

Comment (pleerock) [reply src/util/OrmUtils.ts]: ah sorry looks like equals in in a Function. Why did you remove previous line?

```
if (x.equals instanceof Function && x.equals(y))
```

?

Comment (yoshiaki-natsume) [reply src/util/OrmUtils.ts]: Because it seems that line is just for Buffer class.
So I fixed that line(not remove).

Comment (pleerock) [reply src/util/OrmUtils.ts]: hmmm I wouldn't assume that. For example mongodb's ObjectId has equals method as well. I don't remember why exactly we need this (my bad on missing comment) but we need to put it back until we figure out that we really don't need it.

Comment (pleerock) [reply src/util/OrmUtils.ts]: Also please add a test that **fails** if someday somebody remove lines of code you have added.

Comment (yoshiaki-natsume) [reply src/util/OrmUtils.ts]: yes, I think we need more unit tests to describe logic.
ok, I'll prepare unit test for this case.

**Comment by yoshiaki-natsume** [comment]
added unit test and fix condition

**Comment by yoshiaki-natsume** [comment]
I made related Pull Request #2476

**Comment by yoshiaki-natsume** [comment]
fixed unit tests

File: src/util/OrmUtils.ts [hunk 2]
```diff
@@ -177,6 +177,13 @@ export class OrmUtils {
         if (x === y)
             return true;
 
+        // Check if both argument is Buffer
+        // Buffer.equals Argument must be a Buffer
+        if (Buffer.isBuffer(x) && Buffer.isBuffer(y))
+            return x.equals(y);
+        if (Buffer.isBuffer(x) || Buffer.isBuffer(y))
```
Comment (imnotjames) [inline src/util/OrmUtils.ts:184]: So there is never a case where a buffer could match something like a uintarray?

File: test/functional/util/compare-objects.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        post.id = new Buffer([0]);
+        await connection.manager.save(post);
+        const counters = new Counters();
+        counters.id = new Buffer([0]);
+        await connection.manager.save(counters);
+
+        // then update its properties and save again
+        const loadedPost = await connection.getRepository(Post).findOne();
+        loadedPost!.counters = counters!;
+        await connection.manager.save(loadedPost!);
```
Comment (imnotjames) [inline test/functional/util/compare-objects.ts:30]: So there shouldn't be an exception here?  That's the expected results?

File: test/functional/util/compare-objects.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+import {Connection} from "../../../src/connection/Connection";
+import {Post} from "./entity/Post";
+import {Counters} from "./entity/Counters";
+
+describe("util > compare2Objects functionality", () => {
+
+    let connections: Connection[];
+    before(async () => connections = await createTestingConnections({
+        entities: [Post, Counters],
+        enabledDrivers: ["mysql"]
```
Comment (imnotjames) [inline test/functional/util/compare-objects.ts:12]: Should this be opened up for all drivers?  Or at least, more drivers?

**Review by imnotjames — COMMENTED** [review]
Can you rebase and fix conflicts

**Comment by AlexMesser** [comment]
This is quite old PR, and since there is no community feedback on this issue, I assume there is no interested in this change from the community which makes these changes disputable. Closing for now.

**Pull request closed without merging**